pkg/workspace/assert.go

```diff
@@ -0,0 +1,162 @@
+		return nil, fmt.Errorf("Invalid @%s annotation - expected @%s to have 2-tuple as argument(s), but found no arguments (by %s)", AnnotationAssertValidate, AnnotationAssertValidate, validationPosition.AsCompactString())
+	}
+	for _, arg := range annotation.Args {
+		ruleTuple, ok := arg.(starlark.Tuple)
+		if !ok {
+			return nil, fmt.Errorf("Invalid @%s annotation - expected @%s to have 2-tuple as argument(s), but found: %s (by %s)", AnnotationAssertValidate, AnnotationAssertValidate, arg.String(), validationPosition.AsCompactString())
+		}
+		if len(ruleTuple) != 2 {
+			return nil, fmt.Errorf("Invalid @%s annotation - expected @%s 2-tuple, but found tuple with length %v (by %s)", AnnotationAssertValidate, AnnotationAssertValidate, len(ruleTuple), validationPosition.AsCompactString())
+		}
+		message, ok := ruleTuple[0].(starlark.String)
+		if !ok {
+			return nil, fmt.Errorf("Invalid @%s annotation - expected @%s to have string describing a valid value as the first item in the 2-tuple, but found type: %s (by %s)", AnnotationAssertValidate, AnnotationAssertValidate, ruleTuple[0].Type(), validationPosition.AsCompactString())
+		}
+		lambda, ok := ruleTuple[1].(starlark.Callable)
+		if !ok {
+			return nil, fmt.Errorf("Invalid @%s annotation - expected @%s to have an assertion function as the second item in the 2-tuple, but found type: %s (by %s)", AnnotationAssertValidate, AnnotationAssertValidate, ruleTuple[1].Type(), validationPosition.AsCompactString())
+		}
+		rules = append(rules, Rule{
+			msg:       message.String(),
+			assertion: lambda,
+			position:  validationPosition,
+		})
+	}
+
+	return rules, nil
+}
+
+// A Rule represents an argument to an @assert/validate annotation;
+// it contains a string description of what constitutes a valid value,
+// and a function that asserts the rule against an actual value.
+// One @assert/validate annotation can have multiple Rules.
+type Rule struct {
+	msg       string
+	assertion starlark.Callable
+	position  *filepos.Position
+}
+
+// Validate runs the assertion from the Rule with the node's value as arguments.
+//
+// Returns an error if the assertion returns False (not-None), or assert.fail()s.
+// Otherwise, returns nil.
+func (r Rule) Validate(node yamlmeta.Node, thread *starlark.Thread) error {
+	var nodeValue starlark.Value
+	switch node.(type) {
+	case *yamlmeta.DocumentSet, *yamlmeta.Array, *yamlmeta.Map:
+		panic(fmt.Sprintf("@%s annotation at %s - not supported on %s at %s", AnnotationAssertValidate, r.position.AsCompactString(), yamlmeta.TypeName(node), node.GetPosition().AsCompactString()))
+	default:
+		values := node.GetValues()
+		nodeValue = yamltemplate.NewGoValueWithYAML(values[0]).AsStarlarkValue()
+	}
+
+	result, err := starlark.Call(thread, r.assertion, starlark.Tuple{nodeValue}, []starlark.Tuple{})
+	if err != nil {
+		return fmt.Errorf("%s requires %s; %s (by %s)", node.GetPosition().AsCompactString(), r.msg, err.Error(), r.position.AsCompactString())
+	}
```

Comment:
Because we want to provided a user-friendly identifier/key of the offending value, I think we _need_ to handle all the Node types.

Here's one approach:

```suggestion
func (r Rule) Validate(node yamlmeta.Node, thread *starlark.Thread) error {
	var key string
	var nodeValue starlark.Value
	switch typedNode := node.(type) {
	case *yamlmeta.DocumentSet, *yamlmeta.Array, *yamlmeta.Map:
		panic(fmt.Sprintf("@%s annotation at %s - not supported on %s at %s", AnnotationAssertValidate, r.position.AsCompactString(), yamlmeta.TypeName(node), node.GetPosition().AsCompactString()))
	case *yamlmeta.MapItem:
		key = fmt.Sprintf("%s", typedNode.Key)
		nodeValue = yamltemplate.NewGoValueWithYAML(typedNode.Value).AsStarlarkValue()
	case *yamlmeta.ArrayItem:
		key = "array item"
		nodeValue = yamltemplate.NewGoValueWithYAML(typedNode.Value).AsStarlarkValue()
	case *yamlmeta.Document:
		key = "document"
		nodeValue = yamltemplate.NewGoValueWithYAML(typedNode.Value).AsStarlarkValue()
	}

	result, err := starlark.Call(thread, r.assertion, starlark.Tuple{nodeValue}, []starlark.Tuple{})
	if err != nil {
		return fmt.Errorf("%s (%s) requires %s; %s (by %s)", key, node.GetPosition().AsCompactString(), r.msg, err.Error(), r.position.AsCompactString())
	}

```